### PR TITLE
Wip resnet exploration

### PR DESCRIPTION
**Pull request by NicolasNoya**
Here is my work of today, basically here is some exploration of how to implement the finetuning of a model using Pytorch and the resnet50 recommended by Julián. For my future work I' ll  define an even bigger augmentation for my data, freeze the resnet weights and start with the training of the model locally (with low epochs) so everything is ready to upload and train in the cloud.

### CODE REVIEW

File: research/classification_layer.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        counter = 0
+        layer = (
+            in_features  # Cahnge this to the dimension of the output from the resnet.
+        )
+
+        layers_array = []
+
+        # layers_array.append(nn.Flatten())
+
+        for i in list_of_weights:
```
Comment (j-oflaherty) [inline research/classification_layer.py:45]: Remeber how a for loop works in python. When calling a for on a list, the variable `i` in this case (that we should choose a better name for) takes on the value of `list_of_weights[0]` on the first pass, `list_of_weights[1]` on the second pass, `list_of_weights[n]` on the n+1 pass. So there's no need to use a counter.

If you for need both the index and the item, you can use enumerate:
```python
example_list = ['a', 'b', 'c']
for i, item in example_list:
    print(i, item)

# Out
# 0 a
# 1 b
# 2 c
```

File: research/classification_layer.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+        self.layers = nn.Sequential(*layers_array)
+        self.output = nn.Linear(layer, num_classes)
+
+        # x represents our data
+
+    def forward(self, x):
+        y = x
+        for layer in self.layers:
+            y = layer(y)
```
Comment (j-oflaherty) [inline research/classification_layer.py:64]: Since you packed the layers in a Sequential, you can simply call
```suggestion
        y = self.layers(y)
```

File: research/classification_layer.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+from sklearn.model_selection import train_test_split
+from torch.utils.data import TensorDataset, DataLoader
+import torch.nn.functional as F
+from sklearn.metrics import accuracy_score, f1_score
+import torchvision.transforms as T
+from PIL import Image
+import os
+from numpy import expand_dims
+import pretrainedmodels as pm
+
```
Comment (j-oflaherty) [inline research/classification_layer.py:22]: There are way more imports than modules there are used in the file. When doing research this won't be a problem (just will take longer to start the kernel since it's loading a lot of modules) but keep in mind when writing the actual scripts any unused imports should be removed.

File: models/data_set.py [hunk 4]
```diff
@@ -1,21 +1,24 @@
 import os
-from torchvision.io import read_image
+from torchvision.io import read_image, ImageReadMode
 from torch.utils.data import Dataset
 from torch import Tensor
+from torchvision import transforms
 
+
+# Change the 0 value here for recommendations in the torch page
 label_to_int = {
```
Comment (j-oflaherty) [inline models/data_set.py:9]: It's common practice, and you will see the IDE has a special highliting for it, to name the constants in upper case.

File: models/data_set.py [hunk 5]
```diff
@@ -27,18 +30,18 @@ def __init__(
...
+            "Cleavers": 3,
+            "Common Chickweed": 4,
+            "Common wheat": 5,
+            "Fat Hen": 6,
+            "Loose Silky-bent": 7,
+            "Maize": 8,
+            "Scentless Mayweed": 9,
+            "Shepherds Purse": 10,
+            "Small-flowered Cranesbill": 11,
+            "Sugar beet": 12,
```
Comment (j-oflaherty) [inline models/data_set.py:44]: Usually when writing code, try to define constants in a single place. In this case, we can use the already defined LABAL_TO_INT as the default value:
```Sugestion
label_to_int: Dict[str, int] = LABEL_TO_INT
```

File: models/data_set.py [hunk 6]
```diff
@@ -1,21 +1,24 @@
 import os
-from torchvision.io import read_image
+from torchvision.io import read_image, ImageReadMode
 from torch.utils.data import Dataset
 from torch import Tensor
+from torchvision import transforms
 
+
+# Change the 0 value here for recommendations in the torch page
```
Comment (j-oflaherty) [inline models/data_set.py:8]: Remember that if you leave a comment, it should work in the context it's in. This comment refers to a change in the code you made, so it'll out of context from now on. A better suited comment can be "start at 1 as recommended by torch" and a link to where the info is.

**Review by j-oflaherty — CHANGES_REQUESTED** [review]
Good progress, but there are some corrections to be made. Check out the comments we left regarding the code.

On another note, remember PRs should be on features instead of daily progress. Our next objective will be to create the model. So create a branch called resnet_model or something like that, descriptive of the work you are going to do in that branch. Once you feel the model is completed, create the PR for agus or me to review.

File: models/data_set.py
Comment (a-freitas) [inline models/data_set.py:1]: Commonly used "dataset" instead

File: models/data_set.py [hunk 7]
```diff
@@ -57,11 +60,13 @@ def __init__(
     def __len__(self) -> int:
         return len(self.img_labels)
 
+    # Here I add a resize image function, so the getitem will fix the sizes of an image
+    # in order to be possible to stack them
```
[the remark this answers is not in the document]
Comment (a-freitas) [reply models/data_set.py:64]: Just to add to this discussion, comments should be written in the third person. Try to think of it more as a report or brief clarification for anyone who's looking at the code.
Sometimes it's useful to ask yourself if the logic is clear enough without the comment or not, this way you can keep them to a minimum. A good example where comments are useful is at the beginning of complex functions that the name may not clarify enough what they are doing.
Also, a few word comments can be added at the end of the line if necessary.

File: research/20231003_finetuning_resent_50.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
+#%%
+# Check model.eval
+import research.start
```
Comment (a-freitas) [inline research/20231003_finetuning_resent_50.py:3]: Try to follow import rules from PEP 8 https://peps.python.org/pep-0008/#imports

File: research/classification_layer.py (hunk 2 above)
Comment (NicolasNoya) [reply research/classification_layer.py:64]: Cool I didn´t know this!

**Comment by a-freitas** [comment]
Try keeping the third person and a more concise description of the PR (as discussed with the code).

Also, it's great you are using a PR draft to show that the work is still in progress and describing that with WIP as well. As the name of the branch may be saved in history, avoid including the "WIP" in its name

File: research/classification_layer.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        layers_array = []
+
+        # layers_array.append(nn.Flatten())
+
+        for i in list_of_weights:
+            batchnomr_layer = nn.BatchNorm1d(layer)
+            layers_array.append(batchnomr_layer)
+            linear_layer = nn.Linear(layer, list_of_weights[counter])
+            layers_array.append(linear_layer)
+            activation = nn.ReLU6()
```
[the remark this answers is not in the document]
Comment (NicolasNoya) [reply research/classification_layer.py:50]: I thought it was LRelu, my mistake

File: research/classification_layer.py (hunk 3 above)
Comment (NicolasNoya) [reply research/classification_layer.py:22]: That's correct

File: models/head_classifier.py [hunk 10]
```diff
@@ -2,28 +2,23 @@
 
 
 class ClassifierHead(nn.Module):
-    def __init__(self, in_features, list_of_weights, num_classes=12):
+    def __init__(self, in_features, list_of_neurons, num_classes=12):
```
Comment (j-oflaherty) [inline models/head_classifier.py:5]: Add typing hints. For list_of_neurons use List from typing:

```python
from typing import List

def __init__(self, list_of_neurons: List[int]):
```

File: models/head_classifier.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+from typing import List
+from torch import Tensor
+
+
+class ClassifierHead(nn.Module):
+    def __init__(
+        self, in_features: int, list_of_neurons: List(int), num_classes: int = 12
+    ):
+        super(ClassifierHead, self).__init__()
+        self.sig = nn.Sigmoid()
```
Comment (j-oflaherty) [inline models/head_classifier.py:11]: Move self.sigmoid this to the end, after self.output, to separate the hyper-paramaters from the model layers. This will make it more readable and help you visualize the full model forward from the model declaration.

**Review by j-oflaherty — APPROVED** [review]
Good work integrating all the comments we left.
